rkyv/src/collections/btree_map/mod.rs

```diff
@@ -365,6 +474,77 @@ impl<K, V> ArchivedBTreeMap<K, V> {
     }
 }
 
+#[inline]
+fn key_value_of_node<K, V>(node: (&Node<[LeafNodeEntry<K, V>]>, usize)) -> (&K, &V) {
+    let entry = &node.0.tail[node.1];
+    (&entry.key, &entry.value)
+}
+
+#[inline]
+fn next_node<'a, K, V>(cur: (&'a Node<[LeafNodeEntry<K, V>]>, usize)) -> Option<(&'a Node<[LeafNodeEntry<K, V>]>, usize)> {
+    let mut index = cur.1 + 1;
+    let mut leaf = cur.0;
+    if index == leaf.tail.len() {
+        if leaf.header.ptr.is_null() {
+            return None;
+        }
+        index = 0;
+        leaf = unsafe {
+            let new_leaf: NonNull<NodeHeader> = NonNull::new_unchecked(leaf.header.ptr.as_ptr() as *mut _);
+            new_leaf.as_ref().classify_leaf::<K, V>()
+        };
+    }
+    Some((leaf, index))
+}
+
+#[inline]
+fn prev_node<'a, K, V>(tree: &'a ArchivedBTreeMap<K, V>, cur: (&'a Node<[LeafNodeEntry<K, V>]>, usize)) -> Option<(&'a Node<[LeafNodeEntry<K, V>]>, usize)>
+where K: Ord {
+    if cur.1 > 0 {
+        return Some((cur.0, cur.1 - 1));
+    }
+
+    let k = key_value_of_node(cur).0;
+
+    let mut current = tree.root()?;
+    loop {
+        match current {
+            ClassifiedNode::Inner(inner) => {
+                let scan = inner
+                    .tail
+                    .binary_search_by(|probe| probe.key.borrow().cmp(k));
+
+                let mut next = match scan {
+                    Ok(i) => unsafe { &*inner.tail[i].ptr.as_ptr() },
+                    Err(0) => unsafe { &*inner.header.ptr.as_ptr() },
+                    Err(i) => unsafe { &*inner.tail[i - 1].ptr.as_ptr() }
+                };
+
+                let mut next_classified: ClassifiedNode<'a, K, V> = next.classify();
+                if let ClassifiedNode::Leaf(leaf) = next_classified {
+                    if std::ptr::addr_eq(leaf, cur.0) {
+                        next = match scan {
+                            Ok(0) => unsafe { &*inner.header.ptr.as_ptr() },
```

Comment:
It looks like this case will fail to find the previous entry if the current entry is the first in a subtree. Here's an example:
```
    root
    /  \
left   right
/  \   /  \
a  b  c   d
```
If the current entry is the first one of `b`, then the last inner node before it will be `left`. However, there is no leaf node before `b` since it's also the first leaf node of `left`. The right entry to return is the last entry of `c`.

I could be misunderstanding since the tests are passing, but it's also possible that the test input size needs to be larger for this failure mode to express.

I think the only way to do this without allocating would be to write this recursively: each node finds the exact entry for `current` and then returns the previous entry or "beginning of subtree". If a non-root node returns "beginning of subtree" then the parent gets the last entry of the next branch. If the root returns "beginning of subtree" then there is no previous entry.